### PR TITLE
#5 - Implement login page

### PR DESCRIPTION
**Pull request by tonnypeperonny**
##### TRELLO INFO

N5

##### CHANGE TYPE

- New feature (non-breaking change which adds functionality)

##### DESCRIPTION

Add login/logout
Changed logger files path to Appdata/ReKreator
Create SignIn view

### CODE REVIEW

File: src/ReKreator.Web/Authorization/SignOut/ISignOutOperation.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿namespace ReKreator.Web.Authorization.SignOut
+{
+    public interface ISignOutOperation
+    {
+        void Logout();
```
Comment (a-yandulski) [inline src/ReKreator.Web/Authorization/SignOut/ISignOutOperation.cs]: To be consistent with other operations please rename to **SignOut**

File: src/ReKreator.Web/Authorization/SignOut/SignOutOperation.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        private readonly IAuthenticationManager authManager;
+
+        public SignOutOperation(IAuthenticationManager authManager)
+        {
+            this.authManager = authManager;
+        }
+
+        public void Logout()
+        {
+            logger.Info($"Info: The user \"{HttpContext.Current.User.Identity.Name}\" left their account.");
```
Comment (a-yandulski) [inline src/ReKreator.Web/Authorization/SignOut/SignOutOperation.cs]: Move logging to the bottom of the method

File: src/ReKreator.Web/Controllers/AccountController.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace ReKreator.Web.Controllers
+{
+    public class AccountController : Controller
+    {
+        private readonly ISignInOperation signInManager;
+        private readonly ISignOutOperation signOutManager;
+        private readonly ISignUpOperation manager;
+
+        public AccountController(ISignInOperation manager, ISignOutOperation signOutManager, ISignUpOperation manager1)
```
Comment (a-yandulski) [inline src/ReKreator.Web/Controllers/AccountController.cs]: Fix parameter name "**manager1**"

File: src/ReKreator.Web/Controllers/AccountController.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+using ReKreator.Web.Helpers;
+using ReKreator.Web.Models;
+
+namespace ReKreator.Web.Controllers
+{
+    public class AccountController : Controller
+    {
+        private readonly ISignInOperation signInManager;
+        private readonly ISignOutOperation signOutManager;
+        private readonly ISignUpOperation manager;
```
Comment (a-yandulski) [inline src/ReKreator.Web/Controllers/AccountController.cs]: Fix field name. Better name is "**signUpOperation**"

File: src/ReKreator.Web/Controllers/AccountController.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+        public AccountController(ISignInOperation manager, ISignOutOperation signOutManager, ISignUpOperation manager1)
+        {
+            this.signInManager = manager;
```
Comment (a-yandulski) [inline src/ReKreator.Web/Controllers/AccountController.cs]: Add null-checks for new parameters

File: src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+namespace ReKreator.Web.Authorization.SignIn
+{
+    public class SignInOperation : ISignInOperation
+    {
+        private static Logger logger = LogManager.GetCurrentClassLogger();
+
+        private readonly UserManager<ApplicationUser> userManager;
+        private readonly IAuthenticationManager authManager;
+
+        public SignInOperation(UserManager<ApplicationUser> userManager,IAuthenticationManager authManager)
```
Comment (a-yandulski) [inline src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs]: Add space between params

File: src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            this.userManager = userManager;
```
Comment (a-yandulski) [inline src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs]: Add null checks

File: src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            this.authManager = authManager;
+        }
+
+        public ClaimsIdentity SignIn(SignInModel model)
+        {
```
Comment (a-yandulski) [inline src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs]: Add null check for **model**

File: src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            ClaimsIdentity claim = null;
+            try
+            {
+                var user = model.Login.Contains("@")
+                    ? userManager.FindByEmail(model.Login)
+                    : userManager.FindByName(model.Login);
+
+                if (user != null && userManager.CheckPassword(user, model.Password))
+                {
+                    claim = userManager.CreateIdentity(user, DefaultAuthenticationTypes.ApplicationCookie);
```
Comment (a-yandulski) [inline src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs]: **authenticationType** parameter (current value DefaultAuthenticationTypes.ApplicationCookie) must be provided from outside of the class. The value must be the same as configured in [OwinConfig](https://github.com/Godel-Mastery-v3/ReKreator/blob/b3b333b7a115ab763a696119f9f58bc51cc83ba1/src/ReKreator.Web/App_Start/OwinConfig.cs#L19)

File: src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+                    authManager.SignOut();
+                    authManager.SignIn(new AuthenticationProperties {IsPersistent = true}, claim);
+                    logger.Info($"Info: user \"{user.Email}\" was authorized.");
+                }
+            }
+            catch (Exception e)
+            {
+                logger.Error($"Exception:{e.Message}");
```
Comment (a-yandulski) [inline src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs]: Use overload with 2 parameters (Exception + Format)

File: src/ReKreator.Web/Models/SignInModel.cs [hunk 11]
```diff
@@ -0,0 +1,15 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace ReKreator.Web.Models
+{
+    public class SignInModel
+    {
+        [Required]
+        [MaxLength(250)]
+        public string Login { get; set; }
```
Comment (a-yandulski) [inline src/ReKreator.Web/Models/SignInModel.cs]: This field can represent both **Email** and **UserName**. However the limit specified is for **Email**. If you take a look at **SignUpModel** you will notice that **UserName** has no length limit. You need to decide whether you limit **UserName** field in the **SignUpModel** or you remove the limitation on the **Login** field in **SignInModel**

File: tests/ReKreator.Web.Tests/AccountControllerTest.cs [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+            //act
+            accountController.SignIn(new SignInModel());
+
+            //assert
+            signInMock.Verify(a => a.SignIn(It.IsAny<SignInModel>()));
+        }
+
+        [Test]
+        public void Login_should_return_correctView_if_SignIn_returns_null()
```
Comment (a-yandulski) [inline tests/ReKreator.Web.Tests/AccountControllerTest.cs]: Please add assertions on result.ViewData.ModelState to verify expected error messages

File: tests/ReKreator.Web.Tests/SignInTest.cs [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+            
+            authManager.Setup(a =>
+                a.SignIn(new AuthenticationProperties {IsPersistent = true}, It.IsAny<ClaimsIdentity>()));
+            authManager.Setup(a =>a.SignOut());
+
+            //act
+            signinManager.SignIn(new SignInModel());
+
+            //verify
+            authManager.Setup(a =>
```
Comment (a-yandulski) [inline tests/ReKreator.Web.Tests/SignInTest.cs]: I guess it should be **Verify** instead of **Setup** here.

**Comment by a-yandulski** [comment]
User name does not appear after login in the right top corner 

image

See description, point 4.

**Comment by a-yandulski** [comment]
Please add tests for **SignOutOperation** class

File: src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+        public OperationResult SignIn(SignInModel model)
+        {
+            if (model == null)
+                throw new ArgumentNullException(nameof(model));
+
+            try
+            {
+                var user = model.Login.Contains("@")
+                    ? userManager.FindByEmailAsync(model.Login).Result
```
Comment (a-yandulski) [inline src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs]: Blocking call. Use synchronous version or make the call async all the way (starting from controller action).

File: src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+                    : userManager.FindByNameAsync(model.Login).Result;
+
+                if (user != null && userManager.CheckPasswordAsync(user, model.Password).Result)
```
Comment (a-yandulski) [inline src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs]: Blocking call. Use synchronous version or make the call async all the way (use async\await starting from controller action).

File: src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+                {
+                    var claim = userManager.CreateIdentityAsync(user, cookie.AuthenticationType).Result;
```
Comment (a-yandulski) [inline src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs]: See comment above

**Comment by a-yandulski** [comment]
Please refactor SignUpOperation.SignUp() to return OperationResult\<IdentityResult\>
https://github.com/Godel-Mastery-v3/ReKreator/blob/3d6982adff3e40c1c542b2ad029ee12b21ac199f/src/ReKreator.Web/Authorization/SignUp/SignUpOperation.cs#L27

File: src/OperationDetails/OperationResult.cs [hunk 17]
```diff
@@ -0,0 +1,18 @@
+﻿namespace OperationDetails
```
Comment (a-yandulski) [inline src/OperationDetails/OperationResult.cs]: Strange name for the project and namespace. Should be for example ReKreator.Common and the namespace can be ReKreator.Common.Operations

File: src/OperationDetails/OperationResult.cs [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public class OperationResult
```
Comment (a-yandulski) [inline src/OperationDetails/OperationResult.cs]: Implement 
```csharp 
public class OperationResult<T> : OperationResult
```
 as discussed

File: src/OperationDetails/OperationResult.cs [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        public bool Succedeed { get; set; }
+        public string Message { get; set; }
+
+        public static OperationResult Succeed(string message = null)
```
Comment (a-yandulski) [inline src/OperationDetails/OperationResult.cs]: Implement methods to produce OperationResult\<T\>
```csharp 
public static OperationResult<T> Succeed<T>(T content, string message = null) {...};

public static OperationResult<T> Fail<T>(T content, string errorMessage) {...};
```

**Comment by a-yandulski** [comment]
Below comment has not been resolved yet:
 - SignInModel.Login allowed length contradicts with SignUpModel.UserName length. This may lead to a user not being able to login with long user name (>250 characters). 

I would suggest to set max length of SignUpModel.UserName and SignInModel.Login to 256 chars as per table definition:

image

File: src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs (hunk 16 above)
Comment (tonnypeperonny) [reply src/ReKreator.Web/Authorization/SignIn/SignInOperation.cs]: image

File: src/ReKreator.Web/Models/SignInModel.cs (hunk 11 above)
Comment (a-yandulski) [reply src/ReKreator.Web/Models/SignInModel.cs]: Sorry, my bad. UserName is 50 chars long according to regex. Please leave as is

File: src/ReKreator.Web/Controllers/AccountController.cs [hunk 20]
```diff
@@ -61,7 +61,7 @@ public ActionResult SignIn(SignInModel model)
         {
             if (ModelState.IsValid)
             {
-                var operationResult = signInManager.SignIn(model);
+                var operationResult = signInManager.SignInAsync(model).Result;
```
Comment (a-yandulski) [inline src/ReKreator.Web/Controllers/AccountController.cs]: Make the action async and await the SignInAsync()

File: src/ReKreator.Web/Ioc/IocContainer.cs [hunk 21]
```diff
@@ -32,7 +32,8 @@ public static IContainer ConfigureContainer()
             builder.RegisterType<SignUpUserValidator>().As<IIdentityValidator<User>>().InstancePerRequest();
             builder.RegisterType<CookieAuthenticationOptions>().AsSelf().SingleInstance();
             builder.RegisterType<OperationResult>().AsSelf().InstancePerRequest();
-            
+            builder.RegisterGeneric(typeof(OperationResult<>)).AsSelf().InstancePerRequest();
```
Comment (a-yandulski) [inline src/ReKreator.Web/Ioc/IocContainer.cs]: What these registrations for?

            builder.RegisterType<OperationResult>().AsSelf().InstancePerRequest();
            builder.RegisterGeneric(typeof(OperationResult<>)).AsSelf().InstancePerRequest();

**Pull request merged**